src/app/component/shared/components/calendar-base/calendar-base.component.ts

```diff
@@ -213,39 +237,167 @@ export class CalendarBaseComponent implements OnInit, OnDestroy {
+      this.getClass = 'enrolled-past-day';
+      return this.getClass;
+    } else if (dayItem.hasHabitsInProgress && dayItem.numberOfDate < (dayItem.date.getDate() - 7) && !dayItem.areHabitsDone) {
+      this.getClass = 'unenrolled-past-day';
+      return this.getClass;
+    } else if (dayItem.hasHabitsInProgress && dayItem.areHabitsDone) {
+      this.getClass = 'enrolled-day';
+      return this.getClass;
+    } else if (dayItem.hasHabitsInProgress && !dayItem.areHabitsDone) {
+      this.getClass = 'unenrolled-day';
+      return this.getClass;
     }
   }
 
-  public markCalendarDays(isMonthCalendar, days) {
-    days.forEach(day => {
-      const date = this.formatDate(isMonthCalendar, day);
-      if (new Date().setHours(0, 0, 0, 0) >= new Date(date).setHours(0, 0, 0, 0)) {
-        this.habitAssignService.getHabitAssignByDate(date).pipe(
-          take(1)
-        ).subscribe((habits: HabitAssignInterface[]) => {
-          day.hasHabitsInProgress = habits.length > 0;
-          day.areHabitsDone = habits.every((habit: HabitAssignInterface) => {
-            return habit.habitStatusCalendarDtoList.some((habitEnrollDate: HabitStatusCalendarListInterface) => {
-              if (habitEnrollDate.enrollDate === date) {
-                return true;
-              }
-              return false;
-            });
-          });
-        });
+  checkHabitListEditable(isMonthCalendar, dayItem: CalendarInterface) {
+    if (isMonthCalendar) {
+      this.selectedDay = new Date(dayItem.year, dayItem.month, +dayItem.numberOfDate);
+    } else {
+      this.selectedDay = dayItem.date;
+    }
+    this.isHabitListEditable = false;
+    const currentDate: Date = new Date();
+    if (currentDate.setHours(0, 0, 0, 0) - this.daysCanEditHabits * 24 * 60 * 60 * 1000 >=
+      new Date(this.selectedDay).setHours(0, 0, 0, 0)) {
+      this.isHabitListEditable = true;
+    }
+  }
+
+  checkCanOpenPopup(dayItem: CalendarInterface) {
+    if (dayItem.hasHabitsInProgress) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  openDialogDayHabits(e, isMonthCalendar, dayItem: CalendarInterface) {
+    this.checkHabitListEditable(isMonthCalendar, dayItem);
+    const date = this.formatDate(isMonthCalendar, dayItem);
+    const habits = this.getHabitsForDay(this.userHabitsList, date);
+    const pos = e.target.getBoundingClientRect();
+    const dialogConfig = new MatDialogConfig();
+
+    dialogConfig.disableClose = false;
+    dialogConfig.autoFocus = true;
+    dialogConfig.backdropClass = 'backdropBackground';
+    dialogConfig.position = {
+      top: (pos.y + 20) + 'px',
+      left: (pos.x - 300) + 'px'
+    };
+    dialogConfig.data = {
+      habitsCalendarSelectedDate: this.formatSelectedDate(isMonthCalendar, dayItem),
+      isHabitListEditable: this.isHabitListEditable,
+      habits: habits.habitAssigns
+    };
+
+    const dialogRef = this.dialog.open(HabitsPopupComponent, dialogConfig);
```

Comment:
move const to top of scope

Reply:
sorry didn't see it.
fixed.

Reply:
sorry, not fixed. because I can't open dialog before definition 'dialogConfig'